src/transport/udp.rs

```diff
@@ -332,4 +332,73 @@ impl<'a> UdpHeaderSlice<'a> {
             checksum: self.checksum()
         }
     }
+
+    /// Calculates the upd header checksum based on a ipv4 header.
+    pub fn calc_checksum_ipv4(&self, ip_header: &Ipv4HeaderSlice, payload: &[u8]) -> Result<u16, ValueError> {
+        self.calc_checksum_ipv4_raw(ip_header.source(), ip_header.destination(), payload)
+    }
+
+    /// Calculates the upd header checksum based on a ipv4 header.
+    pub fn calc_checksum_ipv4_raw(&self, source: [u8;4], destination: [u8;4], payload: &[u8]) -> Result<u16, ValueError> {
```

Comment:
It's a little weird to me that `payload` gets passed as a reference, but ownership is taken of `source` and `destination`. I don't think this is wrong, it just looks weird.

Reply:
It's a direct copy of https://github.com/fortanix/etherparse/blob/cab1b5b1d894417fb079acb3daa3f126f2febe2a/src/transport/tcp.rs#L428. This approach is used by the original author of the crate.